File: python/packages/botas/src/botas/tracer_provider.py
"""Lazy OpenTelemetry tracer provider for botas instrumentation.

Returns ``None`` when ``opentelemetry-api`` is not installed — all span
operations become no-ops and the library works without telemetry.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from opentelemetry.trace import Tracer

_tracer: Tracer | None = None
_initialized: bool = False


def get_tracer() -> Tracer | None:
    """Return the shared OpenTelemetry tracer, or ``None`` if unavailable.

    If ``_tracer`` is set (e.g. for testing), returns it directly.
    When ``_initialized`` is True and ``_tracer`` is None, returns None (no-op).
    Otherwise calls ``trace.get_tracer()`` each time to pick up the
    currently configured global TracerProvider.
    """
    if _initialized:
        return _tracer
    try:
        from opentelemetry import trace

        from botas._version import __version__

        return trace.get_tracer("botas", __version__)
    except ImportError:
        return None

File: node/packages/botas-core/src/tracer-provider.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import type { Span, Tracer } from '@opentelemetry/api'
import { createRequire } from 'node:module'

const require = createRequire(import.meta.url)

let _tracer: Tracer | null | undefined // undefined = not yet initialized
let _api: typeof import('@opentelemetry/api') | null = null

/**
 * Returns the shared OpenTelemetry tracer for botas instrumentation.
 * Returns `null` when `@opentelemetry/api` is not installed — all span
 * operations become no-ops and the library works without telemetry.
 */
/** @internal Reset the cached tracer — for testing only. */
export function resetTracer (): void {
  _tracer = undefined
  _api = null
}

export function getTracer (): Tracer | null {
  if (_tracer !== undefined) return _tracer
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    _api = require('@opentelemetry/api') as typeof import('@opentelemetry/api')
    const pkg = require('../package.json') as { version: string }
    _tracer = _api.trace.getTracer('botas', pkg.version)
  } catch {
    _tracer = null
  }
  return _tracer
}

/**
 * Start a named span, set it as the active span in context, and run the callback.
 * Child spans created inside `fn` will be linked as children of this span.
 * When `@opentelemetry/api` is not available, `fn` is called directly with a `null` span.
 */
export async function withActiveSpan<T> (
  name: string,
  fn: (span: Span | null) => Promise<T>
): Promise<T> {
  const tracer = getTracer()
  if (!tracer || !_api) return fn(null)

  const span = tracer.startSpan(name)
  const ctx = _api.trace.setSpan(_api.context.active(), span)
  return _api.context.with(ctx, async () => {
    try {
      return await fn(span)
    } catch (err) {
      span.recordException(err instanceof Error ? err : new Error(String(err)))
      throw err
    } finally {
      span.end()
    }
  })
}
